sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunner.scala

```diff
@@ -94,8 +88,18 @@ class ArrowPythonRunner(
             writer.writeBatch()
             arrowWriter.reset()
           }
-        } {
           writer.end()
```

Comment:
This is moved out of the finally block because this function will throw exception if the thread is interrupted. writer.end() also doesn't do any clean up - it just writes some final bits to the output channel so we don't need to call it in clean up.

Reply:
So basically your change makes sure that `root` and `allocator` is safely closed in the final block and hence we don't need the safe belt in `TaskCompletionListener`?

Reply:
The routine in `TaskCompletionListener` is not really a safe belt, it actually causes the race condition. But you are right, I believe putting the cleanup routine in finally block is enough.

Reply:
maybe add a comment to explain the same - that this should be outside of the finally?